source/validate_builtins.cpp

```diff
@@ -0,0 +1,1474 @@
+  auto it = std::find_if(
+      decorations.begin(), decorations.end(), [](const Decoration& d) {
+        return SpvDecorationBuiltIn == d.dec_type();
+      });
+  assert(it != decorations.end());
+  return it->params()[0];
+}
+
+// Returns the Storage class of the instruction
+uint32_t getStorageClass(Instruction inst) {
+  switch (inst.opcode()) {
+    case SpvOpTypePointer: return inst.words()[2];
+    case SpvOpTypeForwardPointer: return inst.words()[2];
+    case SpvOpVariable: return inst.words()[3];
+    case SpvOpGenericCastToPtrExplicit: return inst.words()[4];
+    default: return SpvStorageClassMax;
+  }
+}
+
+// TODO(jcaraban): how to find the entry_points in a more direct way?
+// Returns the execution models of the entry points linked to the builtin
+std::vector<uint32_t> getExecutionModels(
+      ValidationState_t& vstate, Instruction inst) {
+  std::vector<uint32_t> exec_vec;
+  // Walks the entry functions in the module
+  for (auto entry_func : vstate.entry_points()) {
+    const auto &interfaces = vstate.entry_point_interfaces(entry_func);
+    // Filtering those interfacing the built-in of interest
+    if (isIncludedIn(inst.id(),interfaces)) {
+      auto func_inst = vstate.FindDef(entry_func);
+      assert(func_inst);
+      // The entry function uses link to the entry points
+      for (auto pair : func_inst->uses()) {
+        auto entry_point = pair.first;
+        if (entry_point->opcode() == SpvOpEntryPoint) {
+          // The entry points contain the execution model
+          auto exec_model = entry_point->words()[1];
+          exec_vec.push_back(exec_model);
+        }
+      }
+    }
+  }
+  return exec_vec;
+}
+
+// Ensures the excution model is included in the list of compatible models
+spv_result_t checkExecutionModel(ValidationState_t& vstate,
+      uint32_t exec_model, std::vector<uint32_t> models) {
+  if (not isIncludedIn(exec_model,models)) {
+    return vstate.diag(SPV_ERROR_INVALID_ID)
+        << "Built-in variables are restricted to only certain"
+           "EXECUTION MODELS (see e.g. Vulkan specification)";
+  }
+  return SPV_SUCCESS;
+}
+
+// Ensures the pair 'exec_model','storage_class' are compatible
+spv_result_t checkStorageClass(ValidationState_t& vstate,
+      uint32_t exec_model, uint32_t storage_class,
+      std::vector<uint32_t> models, std::vector<uint32_t> storages)
+{
+  // If 'exec_model' is included the list of 'models'
+  if (isIncludedIn(exec_model,models)) {
+    // But 'storage_class' is not one of the compatible 'storages'
+    if (not isIncludedIn(storage_class,storages)) {
+      return vstate.diag(SPV_ERROR_INVALID_ID)
+          << "Built-in variables must match the STORAGE CLASS "
+             "of the built-in (see e.g. Vulkan specification)";
+    }
+  }
+  return SPV_SUCCESS;
+}
+
+// Ensures the variable type matches the built-in type
+spv_result_t checkBuiltInType(ValidationState_t& vstate, Instruction inst,
```

Comment:
Need a better function comment.
What are comp_type and comp_size?

Reply:
sure